### ISSUE TITLE
Help needed create and use documents (using Methods) while offline

### ISSUE THREAD

**Issue by wildhart** [jamauro/offline#7]
This package is great for persisting data which already exists for off-line use.

I'm not sure I fully understand the process of queuing methods which create new documents or update existing documents while offline.

Here's my scenario - I'm developing a Site Inspection app.  The inspections are likely to be in remote areas with no mobile coverage.  So the PWA needs to be mobile first. Including loading while completley offline, and getting closed and opened again still while offline (iOS and Android can be fairly aggressive about closing background apps when other apps are open - e.g. the camera).  So far, thanks to this package and my own service worker - I've got all the UI, users and Sites collection cached while offline.

While offline, the user needs to be able to :
* Start a new Inspection (selecting from the list of Sites and Users present)
* Add Photos with notes to the inspection.  These will be geotagged and timestaped.
* Update notes on new and existing photos.

(don't worry about caching photo uploads to S3 storage while offline - I can handle that).

I'm not using `jam:method`, instead I'm using my own method wrapper for strong typing method calls.

I have methods:
* `"inspections.insert"` which takes an `Inspection` document and inserts it into the `inspections` collection and returns the id.
* `"inspectionPhotos.insert"` which takes and `InspectionPhoto` document and adds it ot the `inspectionPhotos` collection.
* `"inspectionPhotos.update"` which modifes an `InspectionPhoto` document.

Note that to simplify syncing think it's going to be easier to store the photos in their own collection with an `inspectionId`, instead of as an arrary on the `Inspection` document.

How does `jam:offline` help me with this?

Ideally I need the client-side method simulation to proceed as normal even if off-line, and for the new/updated document to be in the offline cache for minimongo to use and the UI to react to changes in.  If the server method fails, then the method call should be cached for re-trying when back on-line.  Multiple method calls which update the same document need to be replayed in sequence, or earlier ones ignored.

Is this something `jam:offline` can help with?  Or do I need to keep my own offline cache of inserts/updates of documents made while offline, make sure the UI is always using the latest version, and call the methods to sync to the server only when online?

**Comment by jamauro** [comment]
The [`queueMethod`](https://github.com/jamauro/offline#queuing-methods-when-offline) function as part of this package is what you're looking for. It's used to replay the methods sequentially when the user comes back online. If any of them fail, you can catch them in the `replayErrors` as part of [`handleSyncErrors`](https://github.com/jamauro/offline?tab=readme-ov-file#configuring-optional) and handle them how you'd like, e.g. add them back to the queue, inform the user, etc.

You'll want to make sure your methods are available on the client so you can get the simulation to proceed as normal.

I'm not sure what your method wrapper looks like but assuming it's wrapping `Meteor.applyAsync`, you could add this bit of logic to it:
```js
if (Meteor.isClient) {
  if (!Meteor.status().connected) {
    queueMethod(name, ...args).catch(error => console.error('queueMethod error', error)); // when offline, queue the method with jam:offline
    applyOptions.noRetry = true; // jam:offline will handle retries so we don't want to rely on Meteor's internal retry mechanism
  }
}
    
// invoke the method here with Meteor.applyAsync(...)
```

As an aside, if you have any interest in improving `jam:method` to fit your needs, let me know.

**Comment by wildhart** [comment]
I tried the `queueMethod()` function, thanks.  Previously I didn't realise that it was making the new document available offline - that's brilliant, well done!

However, I've encountered a small issue with it - I think I'm doing something wrong.  When going back on-line I'm getting this error `"Exception while simulating the effect of invoking 'inspections.create' Error: Called saveOriginals twice without retrieveOriginals"`.  Here's some steps to reproduce the problem I'm having:

1. ensure app is online, then refresh.
2.  Go off-line (I'm using my phone connected by USB debugging so I can see the console logs, and I turn off Wifi & Mobile data).  Wait for red-cloud icon in my UI to confirm `Meteor.status().connected === false`
3. call the `inspections.create` offline method wrapper (see below).
4. The method is queued, the inspection appears in my reactive UI - all good! Logs:
```
> offline, queuing method... inspections.create
> Created inspection (2) [null, 'MEDuGsjS9t4rBJN8z']
```
5. Go back online.  The new Inspection disappears - Console message:
```
Exception while simulating the effect of invoking 'inspections.create' Error: Called saveOriginals twice without retrieveOriginals
    at LocalCollection.saveOriginals (local_collection.js:346:13)
    at Object.saveOriginals (collection.js:299:26)
    at store.<computed> [as saveOriginals] (livedata_connection.js:312:38)
    at livedata_connection.js:950:13
    at Array.forEach (<anonymous>)
    at Connection._saveOriginals (livedata_connection.js:949:33)
    at Connection.apply (livedata_connection.js:646:14)
    at Collection._callMutatorMethod (allow-deny.js:442:27)
    at Collection.insert (collection.js:618:27)
    at proto.<computed> [as insert] (hooks.js:96:16)
```
6. The missing inspection exists in my server db, but is NOT in the local indexedDB.  Refreshing the page causes the missing inspection to reapear after syncing from the server.

Here's my generic typed method wrapper:

```ts
import { Random } from 'meteor/random';

export default function TypedMethod<A extends any[], R>(name: string, methodFn?: (this: Meteor.MethodThisType, ...args: A) => R) {

    function fn(...argsAndCallback: [...args: A, callback?: (error: Meteor.Error | null, result: R | null) => void]) {
        return Meteor.call(name, ...argsAndCallback)
    }

...

    fn.offline = (...args: A) => {
        return new Promise<[Meteor.Error | null, R | null]>((resolve) => {
            let noRetry = false;
            args[0]._id = Random.id();  // <- more on this later...
            if (Meteor.isClient) {
                if (!Meteor.status().connected) {
                    // when offline, queue the method with jam:offline
                    console.log('offline, queuing method...', name);
                    queueMethod(name, ...args).then(_ => {
                        resolve([null, args[0]._id]);
                    }).catch(e => {
                        console.error('queuedMethodError', e);
                    });
                    noRetry = true;
                }
            }
            Meteor.applyAsync(name, args, { noRetry }, (err, res) => {
                console.log('applyAsync callback', err, res);
                resolve([err, res]);
            });
        });
    }

    return fn;
}
```
This is my `create` method, defined in the client and server:
```ts
export const siteInspectionMethods = {
    ....
    createInspection: TypedMethod('inspections.create', function (inspection: Mongo.OptionalId<SiteInspection>) {
        /************************** VALIDATE PERMISSIONS ***********************************/
        check(this.userId, Match.isStaff);
        /************************** VALIDATE INPUTS ****************************************/
        check(inspection.projectId, String);
        ....
        /***********************************************************************************/

        const newId = inspectionsCollection.insert(inspection);
        return newId;
    }),
};
```
Then to create a new inspection:
```ts
const [err, id] = await siteInspectionMethods.createInspection.offline(this.inspection);
console.log('Created inspection', err, id);
```

What am I doing wrong?

Also, you'll see I've got the line `args[0]._id = Random.id();`.  If I don't include that then the Inspection gets created with a different `_id` when the app goes back online and syncs with the server.  Is that the expected behaviour?

**Comment by jamauro** [comment]
Ah, I forgot you also need to set another flag in the options for `Meteor.applyAsync`

```js
const options = {
  ...(Meteor.isFibersDisabled ? { returnServerResultPromise: true } : { isFromCallAsync: true })
};
```

> gets created with a different _id when the app goes back online and syncs with the server. Is that the expected behaviour?

Yes it's the expected behavior. For inserts, when the method is replayed there will naturally be a new `_id` generated. The doc that was inserted via client-side simulation while offline is then automatically removed during the syncing process.

**Comment by wildhart** [comment]
I'm afraid that doesn't help.  In my case `Meteor.isFibersDisabled === undefined`, and setting `isFromCallAsync: true` causes the `Called saveOriginals twice...` error to occur even when offline.  FYI I'm on Meteor 2.16.  I've tried all combination of both flags true/false but none work.

**Comment by jamauro** [comment]
Hmm, I have seen that error before with 2.16 specifically and Meteor.applyAsync. You can try copying the `DDP._CurrentMethodInvocation` bits from `jam:method` here: https://github.com/jamauro/method/blob/43297c4b81795f3fcbf3b195a4bf6fa16db1b29e/method.js#L402

Note: `Meteor.isFibersDisabled` is for Meteor 3.x purposes.

**Comment by wildhart** [comment]
Hmm, I'm not sure which bits to copy, I'll have a closer look tomorrow.

Firtst though, I thought I'd try `jam:method` just for this method, since supposedly it works automatically with `jam:offline`.  I created the method:
```ts
export const siteInspectionMethods = {
    ...
    upsertJam: createMethod({
        name: 'inspections.upsertJam',
        validate(inspection: SiteInspection) {
            console.log('validate', this.userId, inspection);
            /************************** VALIDATE PERMISSIONS ***********************************/
            check(this.userId, Match.isStaff);
            /************************** VALIDATE INPUTS ****************************************/
            check(inspection._id, String);
            // TODO...
            /***********************************************************************************/
        },
        async run(inspection: SiteInspection) {
            console.log('run', this.userId, inspection);
            await inspectionsCollection.upsertAsync(inspection._id, inspection);
            return inspection._id;
        },
    }),
};
```
Then I call it:
```ts
const id = await siteInspectionMethods.upsertJam(this.inspection);
console.log('Created inspection', id);
```
Observations:
* When offline, this promise never resolves until the app is back online - is that correct?  How does my client code know it can proceed?
* I don't get the error `Error: Called saveOriginals twice without...`
* But, when the app is back online, the newly inserted document disapears from the view, until I refresh the page to get a fresh copy from the database.

So still the same as before, but without the error message...

**Comment by jamauro** [comment]
> When offline, this promise never resolves until the app is back online - is that correct?

Yes, since we're offline, we won't receive the final result from the server. The rest of the client code outside of this promise will run normally.

Hmm, I'm having a hard time repro-ing when I use `jam:method`. I even tried using a mobile device connected via USB. Can you repro it in the browser by doing the following:

1. clear your Application storage. in Chrome, open the browser tools, go to Application -> Storage (on the left) -> Clear site data
2. start your app from the terminal
3. insert a site inspection (just to be sure all looks good when you're online)
4. in the browser console, disconnect the app `Meteor.disconnect()`
5. insert a new site inspection
6. in the browser console. reconnect the app `Meteor.reconnect()`

**Comment by wildhart** [comment]
If I follow your steps I get the same problem.

I think it's something to do setting the `inspection._id = Random.id()` before calling the method.  I was doing this because I need to know what the `_id` is so I can link photos to it while offline.

If I change my method code to an `insertAsync` with no `_id` then it works OK:
```ts
export const siteInspectionMethods = {

    insertJam: createMethod({
        name: 'inspections.insertJam',
        validate(inspection: SiteInspection) {
            /************************** VALIDATE ...
        },
        async run(inspection: SiteInspection) {
            const id = await inspectionsCollection.insertAsync(inspection);
            console.log('Method: inspections.insertJam - new id', id);
            return id;
        },
    }),
};
```

I also have this in the main client code:
```ts
Tracker.autorun(() => {
    const nInspections = inspectionsCollection.find({}).count();
    console.log('nInspections', nInspections);
});
```
Then it's good, but the `_id` which is created when back online is different:
```
  nInspections 0
> Meteor.disconnect()
< undefined
```
Use the UI to create a new Insepction with this code:
```ts
siteInspectionMethods.insertJam(this.inspection);
```
```
  Method: inspections.insertJam - new id 3EQpoox5892fZKH7W
  nInspections 1
> Meteor.reconnect()
< undefined
  Method: inspections.insertJam - new id hTiHEiuz3vXttkPBB
  nInspections 1
```
There are no errors and the new insepection appears OK in my list of inspections, but the _id has changed.

I would prefer to use an `upsertAsync` with a pre-defined `_id` so that (a) I can use the same method to modify the Inspection while offline, and (b) I can link `InspectionPhotos` to the `Inspection._id`.

So to reproduce the bug, I use this method instead:
```ts
export const siteInspectionMethods = {

    upsertJam: createMethod({
        name: 'inspections.upsertJam',
        validate(inspection: SiteInspection) {
            ...
            check(inspection._id, String);
        },
        async run(inspection: SiteInspection) {
            console.log('Method: inspections.upsertJam', inspection._id);
            await inspectionsCollection.upsertAsync(inspection._id, inspection);
            return inspection._id;
        },
    }),
};
```
Then I get this.
```
  nInspections 0
> Meteor.disconnect()
< undefined
```
```ts
// create the inspection
this.inspection._id = Random.id(); // use a known _id
siteInspectionMethods.upsertJam(this.inspection);
```
```
  Method: inspections.upsertJam SJLuxbfAcQpW2R7ja
  nInspections 1
> Meteor.reconnect()
< undefined
  Method: inspections.upsertJam SJLuxbfAcQpW2R7ja

livedata_connection.js:1417 Uncaught Error: Server sent add for existing id: SJLuxbfAcQpW2R7ja
    at Connection._process_added (livedata_connection.js:1417:15)
    at Connection._processOneDataMessage (livedata_connection.js:1214:12)
    at Meteor.connection._processOneDataMessage (ddp.js:34:41)    <-- this is jam:offline - see below
    at Connection._livedata_data (livedata_connection.js:1271:12)
    at Connection.onMessage (livedata_connection.js:1809:12)
    at browser.js:185:11
    at Array.forEach (<anonymous>)
    at ClientStream.forEachCallback (common.js:30:31)
    at socket.onmessage (browser.js:184:14)
    at r.dispatchEvent (sockjs-1.6.1-min-.js:2:2889)

  nInspections 0

> window.location.reload()
...
  nInspections 0
  nInspections 1   <- received from the subscription.
```
Here's the error:
![image](https://github.com/user-attachments/assets/9d2c5a7c-42da-4710-b5bf-2473f9072a07)

Can you reproduce this when you set `_id = Random.id()` before calling your method?  Is there a way to make that work?

**Comment by wildhart** [comment]
FYI, I think I've worked around this for now by letting `_id` be automatic, and instead using my own `refId` property on the Inspection, and using that to connect Photos to an Inspection.

But if would be good to know if you get the same errors when using a pre-set `_id`, and if you can fix that...

**Comment by jamauro** [comment]
Thanks for the additional info. I was able to repro it. I just published `v0.2.1` which will allow using a preset `_id`.

**Comment by wildhart** [comment]
I can confirm this is working well for me, thanks!
